service-base/src/main/java/org/eclipse/hono/service/plan/PrometheusBasedResourceLimitChecks.java

```diff
@@ -261,19 +342,83 @@ private Long extractLongValue(final JsonObject response) {
-        final long inclusiveDaysBetween = DAYS.between(effectiveSince, Instant.now()) + 1;
-        if (inclusiveDaysBetween > 0 && periodInDays > 0) {
-            final long dataUsagePeriodInDays = inclusiveDaysBetween % periodInDays;
-            return dataUsagePeriodInDays == 0 ? periodInDays : dataUsagePeriodInDays;
+    private long calculateDataVolume(final OffsetDateTime effectiveSince, final PeriodMode mode,
+            final long maxBytes) {
+        if (PeriodMode.MONTHLY.equals(mode)
+                && maxBytes > 0
+                && !OffsetDateTime.now(effectiveSince.getOffset()).isBefore(effectiveSince)
+                && YearMonth.from(OffsetDateTime.now(effectiveSince.getOffset())).equals(YearMonth.from(effectiveSince))
+                && effectiveSince.getDayOfMonth() != 1) {
+            final OffsetDateTime lastDayOfMonth = effectiveSince.with(TemporalAdjusters.lastDayOfMonth());
+            final long daysBetween = DAYS
+                    .between(effectiveSince, lastDayOfMonth) + 1;
+            return Double.valueOf(Math.ceil(daysBetween * maxBytes / lastDayOfMonth.getDayOfMonth())).longValue();
         }
-        return -1L;
+        return maxBytes;
+    }
+
+    /**
+     * Calculates the period for which the data usage is to be retrieved from the
+     * prometheus server based on the mode defined by 
+     * {@link TenantConstants#FIELD_PERIOD_MODE}.
+     *
+     * @param effectiveSince The point of time on which the data volume limit came
+     *                      into effect.
+     * @param mode The mode of the period defined by
+     *                   {@link TenantConstants#FIELD_PERIOD_MODE}.
+     * @param periodInDays The number of days defined by 
+     *                     {@link TenantConstants#FIELD_PERIOD_NO_OF_DAYS}. 
+     * @return The period in days for which the data usage is to be calculated.
+     */
+    private long calculateDataUsagePeriod(final OffsetDateTime effectiveSince, final PeriodMode mode,
+            final long periodInDays) {
+        final long inclusiveDaysBetween = DAYS.between(effectiveSince,
+                OffsetDateTime.now(effectiveSince.getOffset())) + 1;
+        switch (mode) {
+        case DAYS:
+            if (inclusiveDaysBetween > 0 && periodInDays > 0) {
+                final long dataUsagePeriodInDays = inclusiveDaysBetween % periodInDays;
+                return dataUsagePeriodInDays == 0 ? periodInDays : dataUsagePeriodInDays;
+            }
+            return 0L;
+        case MONTHLY:
+            if (YearMonth.now().equals(YearMonth.from(effectiveSince)) && effectiveSince.getDayOfMonth() != 1) {
+                return inclusiveDaysBetween;
+            }
+            return OffsetDateTime.now(effectiveSince.getOffset()).getDayOfMonth();
+        default:
+            return 0L;
+        }
+    }
+
+    private long getAndStoreInCache(final ExpiringValueCache<Object, Object> cache, final String key,
```

Comment:
How about naming the method `getOrAddToCache` instead?

Reply:
I also had that option in mind and somehow ended up with the current one with the intention of _get/calculate_ the value and store in the cache. `getOrAddToCache` also sound good.